### ISSUE TITLE
Getting vertex after update property sometimes return in inconsistent value

### ISSUE THREAD

**Issue by anvie** [thinkaurelius/titan#151]
Example:

``` java

Vertex v = g.getVertex(4);
v.setProperty("age", 25);
g.stopTransaction(Conclusion.SUCCESS);

```

After that operation when I'm trying to getting the vertex with id 4 property `age` returning inconsistent value, sometimes got right result sometimes got previous value (before update). Example:

``` java
Vertex v = g.getVertex(4);
int age = v.getProperty("age");  // inconsistent value returned
```

This strange behavior happening in cassandra backend but not in memory storage, and fixed when I'm close the db and reopen it.

I'm looking inside the code and found VertexCache is this because cache not invalidated after updating the vertex properties? 

For your info I'm using multi threaded environment.

**Comment by anvie** [comment]
Oh forgot to mention, this strange behavior happen in 0.2.0 and 0.3.0-SNAPSHOT.

**Comment by mbroecheler** [comment]
Is the latter piece of code (i.e. getting the property) happening in a
different thread than the former?
Note that Cassandra has configurable consistency and is an eventually
consistent system. So, if you have a cassandra cluster and make an update
then immediately read back the value it may be possible that the old value
is returned.

On Tue, Feb 5, 2013 at 10:09 PM, Robin Syihab notifications@github.comwrote:

> Oh forgot to mention, this strange behavior happen in 0.2.0 and
> 0.3.0-SNAPSHOT.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/thinkaurelius/titan/issues/151#issuecomment-13168666.

## 

Matthias Broecheler
http://www.matthiasb.com

**Comment by anvie** [comment]
Yeah, the seconds code happening in different thread but using single shared Titan instance. I'm running single node Cassandra in my local machine, I also waiting after update for a 5 minutes and still returning inconsistent value, but it fixed when I close and reopen instance (i.e using `TitanFactory.open(conf)`)

**Comment by anvie** [comment]
I'm testing with other type data, and it's seems only happening on int or long, not string.

**Comment by mbroecheler** [comment]
The fact that it only happens with certain data types is bizarre. Titan
does not cache properties internally, so this must be something that is
happening at the cassandra level.
Can you send me a snippet of code that reproduces the problem? I will then
investigate the bug from there.
Thank you,
Matthias

On Thu, Feb 7, 2013 at 4:24 AM, Robin Syihab notifications@github.comwrote:

> I'm testing with other type data, and it's seems only happening on int or
> long, not string.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/thinkaurelius/titan/issues/151#issuecomment-13233606.

## 

Matthias Broecheler
http://www.matthiasb.com

**Comment by anvie** [comment]
Hi mbroecheler, i'm revised my test, in previous test i was lucky when i changed data type to string its seems consistent, but today i just got inconsistent for string too.
Okay I will create code for debugging this weird anomaly.
thx.
--robin

**Comment by anvie** [comment]
Here is my testing code, I'm create 2 thread DataReader and DataWriter, where DataWriter make changes to vertex, and DataReader only read properties from vertex, it's seems cache not invalidated on DataReader thread when DataWriter make changes on it own context even both thread using single TitanGraph instance.

https://gist.github.com/anvie/4744394

Output example:

![Screen Shot 2013-02-09 at 2 00 12 PM](https://f.cloud.github.com/assets/146631/141844/f451783c-7286-11e2-861a-4c04885d8fec.png)

When I see on above output, the reader.thread output never changes even thread.writer make changes, maybe it is why i got anomaly behavior, because i'm using Titan on thread pooling environment, so sometimes got from up-to-dated Thread (i.e Thread that modifying the vertex) and sometimes got from other Thread (i.e Thread that not modifying the vertex).

is this a bug? Or something wrong with my code?

thx.
--robin

**Comment by mbroecheler** [comment]
Ah, I see what you are doing, Robin. The reason you are seeing this "caching" behavior is because that's essentially what is happening. But at the transaction level and not the database level.
Titan is a transactional database (https://github.com/thinkaurelius/titan/wiki/Transaction-Handling) and as such each transaction maintains its own temporary "view" of the graph. So, in your data reader, when you first read the vertex, it will read that vertex as it exists at that point in time. From that point on, it will maintain that view of the vertex until the end of the transaction. Any other transaction that updates the vertex later (even just 1 ms) does not exist to that DataReader transaction.
Bottom line: You want to close the transaction in DataReader using g.stopTransaction(Conclusion.SUCCESS) at the end of the while loop before the try.

Also, note that you don't need to explicitly invoke g.startTransaction. Please read https://github.com/thinkaurelius/titan/wiki/Transaction-Handling and in particular the GOTCHA section.

**Comment by anvie** [comment]
Owh i see, thx :+1:
